zebrad/src/components/sync.rs

```diff
@@ -171,62 +164,38 @@ where
         // due to protocol limitations
         self.request_genesis().await?;
 
+        // Distinguishes a restart from a start, so we don't delay when starting
+        // the sync process, but we can keep restart logic in one place.
+        let mut started_once = false;
+
         'sync: loop {
-            // Update metrics for any ready tasks, before wiping state
-            while let Some(Some(rsp)) = self.pending_blocks.next().now_or_never() {
-                match rsp.expect("block download and verify tasks should not panic") {
-                    Ok(hash) => tracing::trace!(?hash, "verified and committed block to state"),
-                    Err((e, _)) => {
-                        tracing::trace!(?e, "sync error before restarting sync, ignoring")
-                    }
-                }
+            if started_once {
+                tracing::info!(timeout = ?SYNC_RESTART_TIMEOUT, "waiting to restart sync");
+                self.prospective_tips = HashSet::new();
+                self.downloads.cancel_all();
+                self.update_metrics();
+                delay_for(SYNC_RESTART_TIMEOUT).await;
+            } else {
+                started_once = true;
             }
-            self.update_metrics();
-
-            // Wipe state from prevous iterations.
-            self.prospective_tips = HashSet::new();
-            self.pending_blocks = Box::pin(FuturesUnordered::new());
-            self.update_metrics();
 
             tracing::info!("starting sync, obtaining new tips");
-            if self.obtain_tips().await.is_err() || self.prospective_tips.is_empty() {
-                self.update_metrics();
-                tracing::warn!("failed to obtain tips, waiting to restart sync");
-                delay_for(SYNC_RESTART_TIMEOUT).await;
+            if let Err(e) = self.obtain_tips().await {
+                tracing::warn!(?e);
```

Comment:
Matching on the error value lets us display the error to the user.  We don't need to check whether prospective_tips is empty, because we'll just fall through to the end of the loop.

In a later PR we should try to change the errors so that they don't display a bug report link -- these errors are likely network errors or transient errors (cc @yaahc)

Reply:
(Let's open a ticket, so we don't forget to make this change.)

Reply:
Yes, I just wanted to cc @yaahc here to flag it in context and have some preliminary discussion first.

Reply:
oh interesting, so we want to be able to mark _all_ errors that pass thru this point so that they don't have bug report urls presented with them. that shouldn't be a difficult feature to add. At this point tho there are a bunch of features in color-eyre being motivated by zebra and some other integrations that I've wanted to write that I couldn't easily express so I think we might want to write our own EyreHandler, I'm thinking in a zebra-eyre crate, that we can use to more tightly control these integrations and provide a staging ground where we can test out ideas before I upstream them into color-eyre.

Reply:
I'm not sure what the easiest way to implement it would be, but if there were a method that could turn on or off the bugreport URL for a specific error instance, I think it would be fine to go through and tack it on to every `tracing::warn!` invocation in the loop.

Having a `zebra-eyre` crate could be good.  My only concern is that once that crate exists, it becomes an easy "accumulation point" for extra code, and in the long run I think it would be good if we didn't have any zebra-specific error handling.  So it would be good to have a plan or focus on how to ensure that the staging ground remains a staging ground.

Reply:
`zebra-eyre` would be specific to error reporting, not error handling, so I think it's fine to have specific rules. The default `std::error::Error` integration will make it possible to get good reporting for errors regardless for anyone using our libraries, and if they want to use `eyre` they could then also install our `zebra-eyre` error reporting hook to get customized reporting that is more aware of the concrete types we use and is able to extract more information that isn't exposed through the current version of the error trait.

Reply:
Cool, that sounds fine.  Would you like to make an issue with that plan?